server/block/brewer.go

```diff
@@ -0,0 +1,246 @@
+package block
+
+import (
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/item/inventory"
+	"github.com/df-mc/dragonfly/server/item/recipe"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/df-mc/dragonfly/server/world/sound"
+	"sync"
+	"time"
+)
+
+// brewer is a struct that may be embedded by blocks that can brew potions, such as brewing stands.
+type brewer struct {
+	mu sync.Mutex
+
+	viewers   map[ContainerViewer]struct{}
+	inventory *inventory.Inventory
+
+	duration   time.Duration
+	fuelAmount int32
+	fuelTotal  int32
+}
+
+// newBrewer creates a new initialised brewer. The inventory is properly initialised.
+func newBrewer() *brewer {
+	b := &brewer{viewers: make(map[ContainerViewer]struct{})}
+	b.inventory = inventory.New(5, func(slot int, _, item item.Stack) {
+		b.mu.Lock()
+		defer b.mu.Unlock()
+		for viewer := range b.viewers {
+			viewer.ViewSlotChange(slot, item)
+		}
+	})
+	return b
+}
+
+// InsertItem ...
+func (b *brewer) InsertItem(h Hopper, pos cube.Pos, w *world.World) bool {
+	for sourceSlot, sourceStack := range h.inventory.Slots() {
+		var slot int
+
+		if sourceStack.Empty() {
+			continue
+		}
+
+		if h.Facing == cube.FaceDown {
+			slot = 0
+		} else if _, ok := sourceStack.Item().(item.BlazePowder); ok {
+			slot = 4
+		} else {
+			for brewingSlot, brewingStack := range b.inventory.Slots() {
+				if brewingSlot == 0 || brewingSlot == 4 || !brewingStack.Empty() {
+					continue
+				}
+
+				_, okPotion := sourceStack.Item().(item.Potion)
+				_, okSplash := sourceStack.Item().(item.SplashPotion)
+				if !okPotion && !okSplash {
+					continue
+				}
```

Comment:
Move this out of the for loop. Also allow lingering potions

Reply:
fixed with latest commit